Makes a SocialSituationBot

vvijay874 opened the pull request:
- Creates scenarios for the user
- Provides feedback and overall rating back to the user

vvijay874 committed:
Test

vvijay874 committed:
Functional Social Situation bot

I used the OpenAI API to create a social security bot that can access previous chats and critique a user's response to a social situation.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+
+
+
+load_dotenv(find_dotenv())
```
nit: You don't need to add find_dotenv() as an arg. 
It's unclear in the dotenv docs, but if you look in the source code (likely accessible in VSCode by ctrl+leftclicking the package or function), you'll see that load_dotenv calls find_dotenv for you.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+
+
+load_dotenv(find_dotenv())
+openai.api_key = os.getenv("API_KEY")
```
You should rename your API key. At minimum, it should be more descriptive, referencing that it's your openai key. Their SDK also adds convenience by auto-reading a key from `OPENAI_API_KEY`. 
```
client = OpenAI(
    # This is the default and can be omitted
    api_key=os.environ.get("OPENAI_API_KEY"),
)
```

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+memory = "AI: In a coffee shop, is it appropriate to yell loudly? User: Yes it is appropriate if you need something very urgently"
+numAsks = 0
+
+# def generateResponse(memory, prompt):
```
I'm also a massive fan of hoarding code by commenting it out, but it's pretty bad practice because its a ton of clutter. You should be using git for this. 

Create a commit with your v1 of generateResponse (e.g. git commit -am "Create a generateResponse function")
Revise the function, (git commit -am "Rename generateResponse -> ask_openai; [functionality changes go here]")

You can then reference your first commit and earlier code using git. 
```
git log
# Find your first commit
git checkout SHA(the random characters next to the commit. 
# Truncated SHA (~4+ characters) will also work
```

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+    for x in range(0, len(words) - 10):
+        editResponse = editResponse + words[x] + " "
+    return editResponse
+
```
You should throw the code below into a block
```
if __name__ == "__main__":
    response = ask_openai()
    ....
```

The difference is that when you import this file, the code won't be run. 
```
python3 socialBot.py
# Code in the if name == main block will be run
```

anotherFile.py
```
from socialBot import ask_openai()
....
```

```
python3 anotherFile.py
# socialBot code won't be run
```

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+            prompt = "Help critique this user's response to the social situation found in what you asked previously. Either tell them to try again or say their response was satisfactory. Explain all of their errors. Here's a transcript of previous responses: " + memory
+            response = openai.chat.completions.create(
+                model="gpt-3.5-turbo", 
+                messages=[
```
messages accepts a `list[dict]` (type hinting notation for a list of dictionaries).
It implicitly accepts message history, with each list item representing a message. Instead of memory being a giant string that you append to, you should make memory a list of strings.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+
+
+def ask_openai():
+    global numAsks
```
You should look into classes. I don't want to give you too much code, but here's an example.

```
class Critiquer:
    def __init__(self):
         self.numAsks = 0
         self.memory = ""
    def ask_openai(self): # self is an implicit argument thats automatically passed when called from an instance of Critiquer. 
          [your code]
          self.numAsks += 1
          self.memory += "foobarbaz"

critiquer_instance = Critiquer()
while critiquer_instance.numAsks < 10:
    critiquer_instance.ask_openai()
```

MichaelDeng03 commented on socialBot.py:
```diff
@@ -0,0 +1,76 @@
+# print(printEditResponse(response))
+print("*********************************************************************************")
+memory = memory + "AI: " + response
+memory += "User: But what if I really need to use the bathroom and there is a huge line?"
```
You can use the `input()` function to allow real user input.

vvijay874 committed:
Updated socialBot.py

The Social Situation creator is now fully functional. It can create social situations and also critique user responses. Memory is also fully functional, with the AI being able to remember the conversation up to that point.

SVPandya committed:
Update socialBot.py

Creates a visual graph for with overall rating progress and provides summaries to a parent/therapist.
